ghost/link-redirects/lib/LinkRedirectsService.js

```diff
@@ -1,21 +1,62 @@
+const crypto = require('crypto');
+const DomainEvents = require('@tryghost/domain-events');
+const RedirectEvent = require('./RedirectEvent');
 const LinkRedirect = require('./LinkRedirect');
 
+/**
+ * @typedef {object} ILinkRedirectRepository
+ * @prop {(url: URL) => Promise<LinkRedirect>} getByURL
+ * @prop {(linkRedirect: LinkRedirect) => Promise<void>} save
+ */
+
 class LinkRedirectsService {
+    /** @type ILinkRedirectRepository */
+    #linkRedirectRepository;
+    /** @type URL */
+    #baseURL;
+
+    /**
+     * @param {object} deps
+     * @param {ILinkRedirectRepository} deps.linkRedirectRepository
+     * @param {object} deps.config
+     * @param {URL} deps.config.baseURL
+     */
+    constructor(deps) {
+        this.#linkRedirectRepository = deps.linkRedirectRepository;
+        if (!deps.config.baseURL.pathname.endsWith('/')) {
+            this.#baseURL = new URL(deps.config.baseURL);
+            this.#baseURL.pathname += '/';
+        } else {
+            this.#baseURL = deps.config.baseURL;
+        }
+        this.handleRequest = this.handleRequest.bind(this);
+    }
+
+    /**
+     * Get a unique slug for a redirect which hasn't already been taken
+     *
+     * @returns {Promise<string>}
+     */
+    async getSlug() {
+        return crypto.randomBytes(4).toString('hex');
+    }
+
     /**
      * @param {URL} to
+     * @param {string} slug
      *
      * @returns {Promise<LinkRedirect>}
      */
-    async addRedirect(to) {
-        const from = new URL(to);
-
-        from.searchParams.set('redirected', 'true'); // Dummy for skateboard
+    async addRedirect(to, slug) {
+        const from = new URL(`r/${slug}`, this.#baseURL);
```

Comment:
This is not blocking at all (we can update this later), but wouldn't it be better if the `/r` prefix was part of the `getSlug`, and the `addRedirect` would take a full path instead?

That way we can also use the `LinkRedirectsService` like this later:
```
// Post changed slug (pseudo code)
const oldUrl = post.getUrl({relative: true});
await Post.edit({slug: newSlug}, {id: post.id});
const newUrl = post.getUrl({relative: false});

redirectsService.addRedirect(new URL(newURL), oldUrl);
``` 
(we an also build a completely different service for that too).

Maybe it would also make more sense to put the `slug` parameter first?

Reply:
tbh I don't know if we even want a `/r` prefix - this is all just placeholder for now!


RE: slug param first - I went second because I was considering having it optional - still not sure about that 🤷